drivers/dac/dac_gd32.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2021 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_dac
+
+#include <errno.h>
+#include <drivers/pinctrl.h>
+#include <drivers/dac.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(dac_gd32, CONFIG_DAC_LOG_LEVEL);
+
+/**
+ * GD32 DAC HAL use different DAC0 interface for 2 or 1 output SoCs.
+ * Unify the DAC0 interface to DAC0_xx.
+ */
+#if DT_INST_PROP(0, num_channels) == 1
+#define DAC_CTL_DEN0 DAC_CTL_DEN
+#define DAC0_R8DH    DAC_R8DH
+#define DAC0_R12DH   DAC_R12DH
+#endif
+
+struct dac_gd32_config {
+	uint32_t reg;
+	uint32_t rcu_periph_clock;
+	const struct pinctrl_dev_config *pcfg;
+	uint32_t num_channels;
+	uint32_t reset_val;
+};
+
+struct dac_gd32_data {
+	uint8_t resolutions[2];
+};
+
+static void dac_gd32_enable(uint8_t dacx)
+{
+	switch (dacx) {
+	case 0:
+		DAC_CTL |= DAC_CTL_DEN0;
+		break;
+	case 1:
+		DAC_CTL |= DAC_CTL_DEN1;
+		break;
+	}
+}
+
+static void dac_gd32_disable(uint8_t dacx)
+{
+	switch (dacx) {
+	case 0:
+		DAC_CTL &= ~DAC_CTL_DEN0;
+		break;
+	case 1:
+		DAC_CTL &= ~DAC_CTL_DEN1;
+		break;
+	}
+}
+
+static void dac_gd32_write(struct dac_gd32_data *data,
+			   uint8_t dacx, uint32_t value)
+{
+	switch (dacx) {
+	case 0:
+		if (data->resolutions[dacx] == 8U) {
+			DAC0_R8DH = value;
+		} else {
+			DAC0_R12DH = value;
+		}
+		break;
+	case 1:
+		if (data->resolutions[dacx]  == 8U) {
+			DAC1_R8DH = value;
+		} else {
+			DAC1_R12DH = value;
+		}
+		break;
+	}
+}
+
+static int dac_gd32_channel_setup(const struct device *dev,
+				  const struct dac_channel_cfg *channel_cfg)
+{
+	struct dac_gd32_data *data = dev->data;
+	const struct dac_gd32_config *config = dev->config;
+	uint8_t dacx = channel_cfg->channel_id;
+
+	if (dacx >= config->num_channels) {
+		return -ENOTSUP;
+	}
+
+	/* GD32 DAC only support 8 or 12 bits resolution */
+	if (channel_cfg->resolution != 8U &&
+	    channel_cfg->resolution != 12U) {
```

Comment:
Are you sure that it won't fit as one line?

Reply:
One line is OK, but I think two lines looks better.